File: mahjong/tiles.py
from __future__ import annotations

from dataclasses import dataclass


SUITS = ("m", "p", "s")
HONORS = (
    "\u6771",
    "\u5357",
    "\u897f",
    "\u5317",
    "\u767d",
    "\u767c",
    "\u4e2d",
)
TILE_NAMES = tuple(
    [f"{rank}{suit}" for suit in SUITS for rank in range(1, 10)] + list(HONORS)
)
TILE_TO_INDEX = {tile: index for index, tile in enumerate(TILE_NAMES)}
INDEX_TO_TILE = {index: tile for index, tile in enumerate(TILE_NAMES)}
HONOR_ALIASES = {
    "\u6771": "\u6771",
    "\u5357": "\u5357",
    "\u897f": "\u897f",
    "\u5317": "\u5317",
    "\u767d": "\u767d",
    "\u767c": "\u767c",
    "\u767a": "\u767c",
    "\u4e2d": "\u4e2d",
}


@dataclass(frozen=True)
class ParsedHand:
    counts: tuple[int, ...]
    tiles: tuple[str, ...]


def empty_counts() -> list[int]:
    return [0] * 34


def tile_index(tile: str) -> int:
    normalized = HONOR_ALIASES.get(tile, tile)
    if normalized not in TILE_TO_INDEX:
        raise ValueError(f"Unknown tile: {tile}")
    return TILE_TO_INDEX[normalized]


def tile_name(index: int) -> str:
    return INDEX_TO_TILE[index]


def counts_to_tiles(counts: tuple[int, ...] | list[int]) -> list[str]:
    tiles: list[str] = []
    for index, count in enumerate(counts):
        tiles.extend([tile_name(index)] * count)
    return tiles


def format_counts(counts: tuple[int, ...] | list[int]) -> str:
    return " ".join(counts_to_tiles(counts))


def parse_hand(text: str) -> ParsedHand:
    """Parse separated or compact tile notation into 34 tile counts.

    Supported examples:
    - 1m,2m,3m,East honor,Green dragon
    - 123m 456p 789s honors
    - 123m456p789s followed by honor characters
    """
    counts = empty_counts()
    tiles: list[str] = []
    i = 0
    value = text.strip()

    while i < len(value):
        char = value[i]
        if char.isspace() or char in ",\u3001\uff0c/|":
            i += 1
            continue

        if char.isdigit():
            digits: list[str] = []
            while i < len(value) and value[i].isdigit():
                digit = value[i]
                if digit == "0":
                    raise ValueError("Red fives / 0 notation are not supported.")
                if digit < "1" or digit > "9":
                    raise ValueError(f"Invalid tile rank: {digit}")
                digits.append(digit)
                i += 1
            if i >= len(value) or value[i] not in SUITS:
                raise ValueError("Number tiles must be followed by m, p, or s.")
            suit = value[i]
            i += 1
            for digit in digits:
                _add_tile(counts, tiles, f"{digit}{suit}")
            continue

        if char in HONOR_ALIASES:
            _add_tile(counts, tiles, HONOR_ALIASES[char])
            i += 1
            continue

        raise ValueError(f"Cannot parse tile around: {value[i:]}")

    return ParsedHand(tuple(counts), tuple(tiles))


def validate_counts(counts: tuple[int, ...] | list[int], expected_total: int | None = None) -> None:
    if len(counts) != 34:
        raise ValueError("A hand must use 34 tile counts.")
    if any(count < 0 for count in counts):
        raise ValueError("Tile counts cannot be negative.")
    over_limit = [tile_name(index) for index, count in enumerate(counts) if count > 4]
    if over_limit:
        raise ValueError(f"A tile appears more than four times: {', '.join(over_limit)}")
    if expected_total is not None and sum(counts) != expected_total:
        raise ValueError(f"Expected {expected_total} tiles, got {sum(counts)}.")


def _add_tile(counts: list[int], tiles: list[str], tile: str) -> None:
    index = tile_index(tile)
    counts[index] += 1
    if counts[index] > 4:
        raise ValueError(f"A tile appears more than four times: {tile}")
    tiles.append(tile)

File: packages/mahjong-core/src/tiles.ts
export const SUITS = ["m", "p", "s"] as const;
export type Suit = (typeof SUITS)[number];

export const HONORS = ["東", "南", "西", "北", "白", "發", "中"] as const;

export const TILE_NAMES = [
  ...SUITS.flatMap((suit) => Array.from({ length: 9 }, (_, index) => `${index + 1}${suit}`)),
  ...HONORS,
] as const;

export type Tile = (typeof TILE_NAMES)[number];
export type Counts34 = number[];

const TILE_TO_INDEX = new Map<string, number>(TILE_NAMES.map((tile, index) => [tile, index]));
const HONOR_ALIASES = new Map<string, string>([["発", "發"]]);

export function emptyCounts(): Counts34 {
  return Array(34).fill(0);
}

export function tileIndex(tile: string): number {
  const normalized = HONOR_ALIASES.get(tile) ?? tile;
  const index = TILE_TO_INDEX.get(normalized);
  if (index == null) {
    throw new Error(`Unknown tile: ${tile}`);
  }
  return index;
}

export function tileName(index: number): Tile {
  const tile = TILE_NAMES[index];
  if (tile == null) {
    throw new Error(`Unknown tile index: ${index}`);
  }
  return tile;
}

export function countsToTiles(counts: Counts34): Tile[] {
  validateCounts(counts);
  const tiles: Tile[] = [];
  counts.forEach((count, index) => {
    for (let i = 0; i < count; i += 1) {
      tiles.push(tileName(index));
    }
  });
  return tiles;
}

export function parseHand(text: string): Counts34 {
  const counts = emptyCounts();
  let i = 0;
  const value = text.trim();

  while (i < value.length) {
    const char = value[i]!;
    if (/\s/.test(char) || [",", "、", "，", "/", "|"].includes(char)) {
      i += 1;
      continue;
    }

    if (/[0-9]/.test(char)) {
      const digits: string[] = [];
      while (i < value.length && /[0-9]/.test(value[i]!)) {
        const digit = value[i]!;
        if (digit === "0") {
          throw new Error("Red fives / 0 notation are not supported.");
        }
        if (digit < "1" || digit > "9") {
          throw new Error(`Invalid tile rank: ${digit}`);
        }
        digits.push(digit);
        i += 1;
      }
      const suit = value[i] as Suit | undefined;
      if (!suit || !SUITS.includes(suit)) {
        throw new Error("Number tiles must be followed by m, p, or s.");
      }
      i += 1;
      for (const digit of digits) {
        incrementTile(counts, `${digit}${suit}`);
      }
      continue;
    }

    const honor = HONOR_ALIASES.get(char) ?? char;
    if (TILE_TO_INDEX.has(honor)) {
      incrementTile(counts, honor);
      i += 1;
      continue;
    }

    throw new Error(`Cannot parse tile around: ${value.slice(i)}`);
  }

  return counts;
}

export function sortedHandText(counts: Counts34): string {
  return countsToTiles(counts).join(" ");
}

export function validateCounts(counts: Counts34, expectedTotal?: number): void {
  if (counts.length !== 34) {
    throw new Error("A hand must use 34 tile counts.");
  }
  counts.forEach((count, index) => {
    if (count < 0) {
      throw new Error("Tile counts cannot be negative.");
    }
    if (count > 4) {
      throw new Error(`A tile appears more than four times: ${tileName(index)}`);
    }
  });
  if (expectedTotal != null && sumCounts(counts) !== expectedTotal) {
    throw new Error(`Expected ${expectedTotal} tiles, got ${sumCounts(counts)}.`);
  }
}

export function addTile(counts: Counts34, tile: string): Counts34 {
  const next = counts.slice();
  incrementTile(next, tile);
  return next;
}

function incrementTile(counts: Counts34, tile: string): void {
  const index = tileIndex(tile);
  if (counts[index]! >= 4) {
    throw new Error(`${tileName(index)} is already used four times.`);
  }
  counts[index] += 1;
}

export function removeTile(counts: Counts34, tile: string): Counts34 {
  const next = counts.slice();
  const index = tileIndex(tile);
  if (next[index]! <= 0) {
    throw new Error(`${tileName(index)} is not in the hand.`);
  }
  next[index] -= 1;
  return next;
}

export function sumCounts(counts: Counts34): number {
  return counts.reduce((total, count) => total + count, 0);
}
